SharedPtr plus private destructor has compile error since 1.0.167

The new "raw" function that is generated in 1.0.167 and later causes a compile error for my project that wraps C++ code that uses a design pattern of opaque impl types, public wrapper types with private destructors, and shared_ptrs with custom destructors.

I recreated the problem in a small test project.

```cpp
#include <memory>

class Foo {
    private:
        Foo() = default;
        virtual ~Foo() = default;

    struct Deleter {
        void operator()(Foo* p) const {
            delete p;
        }
    };

    public:
        static std::shared_ptr<Foo> create() {
            return std::shared_ptr<Foo>(new Foo(), Deleter());
        }
};

std::shared_ptr<Foo> create_foo() {
    return Foo::create();
}
```

```rust
#[cxx::bridge]
mod ffi {
    unsafe extern "C++" {
        include!("my_header.h");

        type Foo;

        fn create_foo() -> SharedPtr<Foo>;
    }
}

fn main() {
    let _ = ffi::create_foo();
}
```

This compiles and "works" when using cxx-bridge <= 1.0.166. In 1.0.167 there is a compile error about the private destructor on this generated function:

```rust
void cxxbridge1$shared_ptr$Foo$raw(::std::shared_ptr<::Foo> *ptr, ::Foo *raw) noexcept {
  ::new (ptr) ::std::shared_ptr<::Foo>(raw);
}
```

    /usr/include/c++/15.1.1/bits/shared_ptr_base.h:918:15: error: 'virtual Foo::~Foo()' is private within this context

Is there something I can do to continue wrapping this code (the real project is legacy code that is resistant to change), or can there be an attribute to skip generating the "raw" method? Or perhaps an attribute to specify a custom destructor for the SharedPtr?

--- a/gen/src/builtin.rs
+++ b/gen/src/builtin.rs
@@ -32,6 +32,7 @@ pub(crate) struct Builtins<'a> {
     pub is_complete: bool,
     pub destroy: bool,
     pub deleter_if: bool,
+    pub shared_ptr: bool,
     pub alignmax: bool,
     pub content: Content<'a>,
 }
@@ -128,6 +129,12 @@ pub(super) fn write(out: &mut OutFile) {
     if builtin.layout {
         include.type_traits = true;
         include.cstddef = true;
+        builtin.is_complete = true;
+    }
+
+    if builtin.shared_ptr {
+        include.memory = true;
+        include.type_traits = true;
         builtin.is_complete = true;
     }
 
@@ -409,6 +416,34 @@ pub(super) fn write(out: &mut OutFile) {
             out,
             "  template <typename T> void operator()(T *ptr) {{ ptr->~T(); }}",
         );
+        writeln!(out, "}};");
+    }
+
+    if builtin.shared_ptr {
+        out.next_section();
+        writeln!(
+            out,
+            "template <typename T, bool = ::rust::detail::is_complete<T>::value>",
+        );
+        writeln!(out, "struct is_destructible : ::std::false_type {{}};");
+        writeln!(out, "template <typename T>");
+        writeln!(
+            out,
+            "struct is_destructible<T, true> : ::std::is_destructible<T> {{}};",
+        );
+        writeln!(
+            out,
+            "template <typename T, bool = ::rust::is_destructible<T>::value>",
+        );
+        writeln!(out, "struct shared_ptr_if_destructible {{");
+        writeln!(out, "  explicit shared_ptr_if_destructible(T *) {{}}");
+        writeln!(out, "}};");
+        writeln!(out, "template <typename T>");
+        writeln!(
+            out,
+            "struct shared_ptr_if_destructible<T, true> : ::std::shared_ptr<T> {{",
+        );
+        writeln!(out, "  using ::std::shared_ptr<T>::shared_ptr;");
         writeln!(out, "}};");
     }
 

--- a/gen/src/write.rs
+++ b/gen/src/write.rs
@@ -1857,13 +1857,19 @@ fn write_shared_ptr(out: &mut OutFile, key: &NamedImplKey) {
         writeln!(out, "}}");
     }
 
+    out.builtin.shared_ptr = true;
     begin_function_definition(out);
     writeln!(
         out,
-        "void cxxbridge1$shared_ptr${}$raw(::std::shared_ptr<{}> *ptr, {} *raw) noexcept {{",
+        "bool cxxbridge1$shared_ptr${}$raw(::std::shared_ptr<{}> *ptr, {} *raw) noexcept {{",
         instance, inner, inner,
     );
-    writeln!(out, "  ::new (ptr) ::std::shared_ptr<{}>(raw);", inner);
+    writeln!(
+        out,
+        "  ::new (ptr) ::rust::shared_ptr_if_destructible<{}>(raw);",
+        inner,
+    );
+    writeln!(out, "  return ::rust::is_destructible<{}>::value;", inner);
     writeln!(out, "}}");
 
     begin_function_definition(out);

--- a/macro/src/expand.rs
+++ b/macro/src/expand.rs
@@ -14,6 +14,7 @@ use crate::type_id::Crate;
 use crate::{derive, generics};
 use proc_macro2::{Ident, Span, TokenStream};
 use quote::{format_ident, quote, quote_spanned, ToTokens};
+use std::fmt::{self, Display};
 use std::mem;
 use syn::{parse_quote, punctuated, Generics, Lifetime, Result, Token};
 
@@ -1597,6 +1598,7 @@ fn expand_shared_ptr(
     let begin_span = explicit_impl.map_or(key.begin_span, |explicit| explicit.impl_token.span);
     let end_span = explicit_impl.map_or(key.end_span, |explicit| explicit.brace_token.span.join());
     let unsafe_token = format_ident!("unsafe", span = begin_span);
+    let not_destructible_err = format!("{} is not destructible", display_namespaced(resolve.name));
 
     quote_spanned! {end_span=>
         #[automatically_derived]
@@ -1617,10 +1619,10 @@ fn expand_shared_ptr(
             unsafe fn __raw(new: *mut ::cxx::core::ffi::c_void, raw: *mut Self) {
                 #UnsafeExtern extern "C" {
                     #[link_name = #link_raw]
-                    fn __raw(new: *const ::cxx::core::ffi::c_void, raw: *mut ::cxx::core::ffi::c_void);
+                    fn __raw(new: *const ::cxx::core::ffi::c_void, raw: *mut ::cxx::core::ffi::c_void) -> ::cxx::core::primitive::bool;
                 }
-                unsafe {
-                    __raw(new, raw as *mut ::cxx::core::ffi::c_void);
+                if !unsafe { __raw(new, raw as *mut ::cxx::core::ffi::c_void) } {
+                    ::cxx::core::panic!(#not_destructible_err);
                 }
             }
             unsafe fn __clone(this: *const ::cxx::core::ffi::c_void, new: *mut ::cxx::core::ffi::c_void) {
@@ -2000,6 +2002,21 @@ fn expand_extern_return_type(ret: &Option<Type>, types: &Types, proper: bool) ->
     };
     let ty = expand_extern_type(ret, types, proper);
     quote!(-> #ty)
+}
+
+fn display_namespaced(name: &Pair) -> impl Display + '_ {
+    struct Namespaced<'a>(&'a Pair);
+
+    impl<'a> Display for Namespaced<'a> {
+        fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            for segment in &self.0.namespace {
+                write!(formatter, "{segment}::")?;
+            }
+            write!(formatter, "{}", self.0.cxx)
+        }
+    }
+
+    Namespaced(name)
 }
 
 // #UnsafeExtern extern "C" {...}

--- a/src/shared_ptr.rs
+++ b/src/shared_ptr.rs
@@ -95,6 +95,11 @@ where
     /// The resulting shared pointer is **nonempty** regardless of whether the
     /// input pointer is null, but may be either **null** or **nonnull**.
     ///
+    /// # Panics
+    ///
+    /// Panics if `T` is an incomplete type (including `void`) or is not
+    /// destructible.
+    ///
     /// # Safety
     ///
     /// Pointer must either be null or point to a valid instance of T

--- a/tests/ffi/lib.rs
+++ b/tests/ffi/lib.rs
@@ -25,6 +25,13 @@ use std::os::raw::c_char;
 
 #[cxx::bridge(namespace = "tests")]
 pub mod ffi {
+    extern "C++" {
+        include!("tests/ffi/tests.h");
+
+        type Undefined;
+        type Private;
+    }
+
     #[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
     struct Shared {
         z: usize,
@@ -98,8 +105,6 @@ pub mod ffi {
     }
 
     unsafe extern "C++" {
-        include!("tests/ffi/tests.h");
-
         type C;
 
         fn c_return_primitive() -> usize;
@@ -360,6 +365,8 @@ pub mod ffi {
 
     impl Box<Shared> {}
     impl CxxVector<SharedString> {}
+    impl SharedPtr<Undefined> {}
+    impl SharedPtr<Private> {}
 }
 
 mod other {

--- a/tests/ffi/tests.h
+++ b/tests/ffi/tests.h
@@ -34,6 +34,13 @@ public:
 
 namespace tests {
 
+class Undefined;
+
+class Private {
+private:
+  ~Private();
+};
+
 struct R;
 struct Shared;
 struct SharedString;

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -323,6 +323,24 @@ fn test_unique_to_shared_ptr_null() {
 }
 
 #[test]
+fn test_shared_ptr_from_raw() {
+    let shared = unsafe { SharedPtr::<ffi::C>::from_raw(ptr::null_mut()) };
+    assert!(shared.is_null());
+}
+
+#[test]
+#[should_panic = "tests::Undefined is not destructible"]
+fn test_shared_ptr_from_raw_undefined() {
+    unsafe { SharedPtr::<ffi::Undefined>::from_raw(ptr::null_mut()) };
+}
+
+#[test]
+#[should_panic = "tests::Private is not destructible"]
+fn test_shared_ptr_from_raw_private() {
+    unsafe { SharedPtr::<ffi::Private>::from_raw(ptr::null_mut()) };
+}
+
+#[test]
 fn test_c_ns_method_calls() {
     let unique_ptr = ffi2::ns_c_return_unique_ptr_ns();
 

